File: leetcode/leetcode_py/longest-zigzag-path-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
'''
https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree

1. DFS (passing left, right count)
time: o(n)
space: o(n)
'''
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        def dfs(node, l, r):
            if not node: return 0
            return max(l, r, dfs(node.left, r+1, 0), dfs(node.right, 0, l+1))
        
        return dfs(root, 0, 0)
        

File: leetcode/leetcode_cpp/longest-zigzag-path-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
/*
https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree

1. DFS (passing left, right count)
time: o(n)
space: o(n)
*/
class Solution {
public:
    void dfs(TreeNode* node, int l, int r, int& res) {
        if (!node) return;
        res = max(res, max(l, r));
        dfs(node->left, r + 1, 0, res);
        dfs(node->right, 0, l + 1, res);
    }
    int longestZigZag(TreeNode* root) {
        int res = 0;
        dfs(root, 0, 0, res);
        return res;
    }
};